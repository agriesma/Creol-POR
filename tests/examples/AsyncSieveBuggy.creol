/*
 * The asynchronous sieve example.
 */

/*
 * This version is a naive transformation of the synchronous example to
 * communicate asynchronously.  However, it has an error, since it does
 * not take the unorderedness of the communication into account.
 */

interface Sieve
begin
with Any
  op send(in x: Nat; out r: Bool)
end

class Generator
begin
  var n: Int
  var next: Sieve
  op init == next := new Sieve(2); n := 3
  op run == var r: Bool; l!next.send(n); n := n + 1; !run(); await l?; l?(r)
end

class Sieve(p: Int) implements Sieve
begin
  var next: Sieve
  op init == skip
  op run == skip
  op send(in n: Int; out res: Bool) ==
    var d: Int := n / p, r: Int := n - d * p;
    if r = 0
    then res := false
    else if next /= null
         then
	   await next.send(n; res)
         else
	   next := new Sieve(n); res := true
         end
    end
end
