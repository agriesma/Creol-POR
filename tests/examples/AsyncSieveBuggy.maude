load creol-interpreter .
mod PROGRAM is
protecting CREOL-SIMULATOR .
op classes : -> Configuration [ctor] .
eq classes =
< "Generator" : Cl | Inh: (noInh),
  Par: (noVid),
  Att: ("n" |-> null , "next" |-> null),
  Mtds: (< "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "next" ::= new "Sieve" ( int(2) ) ; "n" ::= int(3) ; return ( emp ) > *
    < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "r" |-> null , "l" |-> null , "_" |-> null, Code: "l" ! "next" . "send" ( "n" ) ; "n" ::= "+" ( "n" :: int(1) ) ; ( "label:0" ! "this" . "run" (emp ) )  ; free( "label:0" ) ; ( await ( "l" ?? ) ) ; ( "l" ? ( "r" ) )  ; return ( emp ) >),
  Ocnt: 0 >

< "Sieve" : Cl | Inh: (noInh),
  Par: ("p"),
  Att: ("p" |-> null , "next" |-> null),
  Mtds: (< "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
    < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
    < "send" : Mtdname | Param: "n", Latt: "n" |-> null , "res" |-> null , "label:0" |-> null , "d" |-> null , "r" |-> null , "_" |-> null, Code: "d" ::= "/" ( "n" :: "p" ) ; "r" ::= "-" ( "n" :: "*" ( "d" :: "p" ) ) ; if "=" ( "r" :: int(0) ) th "res" ::= bool(false) el if "/=" ( "next" :: null ) th "label:0" ! "next" . "send" ( "n" ) ; ( await ( "label:0" ?? ) ) ; ( "label:0" ? ( "res" ) )  el "next" ::= new "Sieve" ( "n" ) ; "res" ::= bool(true) fi fi ; return ( "res" ) >),
  Ocnt: 0 >

.
endm
