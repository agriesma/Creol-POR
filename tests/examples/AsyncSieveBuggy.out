==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Generator" : Cl | Inh: noInh, Par: noVid, Att: "n" |-> null, "next" |->
    null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "next" ::= new
    "Sieve"(int(2)) ; "n" ::= int(3) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, "l" |-> null, "label:0"
    |-> null, "r" |-> null, Code: "l" ! "next" . "send"("n") ; "n" ::= "+"("n"
    :: int(1)) ; "label:0" ! "this" . "run"(emp) ; free("label:0") ; await ("l"
    ??) ; "l" ?("r") ; return(emp) >,
  Ocnt: 0 >

< "Sieve" : Cl | Inh: noInh, Par: "p", Att: "next" |-> null, "p" |-> null,
    Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "send" : Mtdname | Param: "n", Latt: "_" |-> null, "d" |-> null, "label:0"
    |-> null, "n" |-> null, "r" |-> null, "res" |-> null, Code: "d" ::= "/"("n"
    :: "p") ; "r" ::= "-"("n" :: "*"("d" :: "p")) ; if "="("r" :: int(0)) th
    "res" ::= bool(false) el if "/="("next" :: null) th "label:0" ! "next" .
    "send"("n") ; await ("label:0" ??) ; "label:0" ?("res") el "next" ::= new
    "Sieve"("n") ; "res" ::= bool(true) fi fi ; return("res") >,
  Ocnt: 0 >
Bye.
